File: python/blog_client.py
"""The Python implementation of the gRPC route guide client."""

from __future__ import print_function
import datetime

import logging
import random
import time
from typing import List, Tuple
from essential_generators import DocumentGenerator

import grpc
import blog_pb2
import blog_pb2_grpc
import blog_resources

gen = DocumentGenerator()


def create_blogs(stub: blog_pb2_grpc.BlogServiceStub) -> List[str]:
    blog_ids: List[str] = []
    for _ in range(3):
        blog_creation = blog_pb2.BlogCreation(
            author=gen.name(), name=gen.sentence())

        blog_response = stub.CreateBlog(blog_creation)
        print("\nBlog with id %s, name %s and author %s, was created." %
              (blog_response.id, blog_response.name, blog_response.author))
        blog_ids.append(blog_response.id)

    return blog_ids


def get_blog_info(stub: blog_pb2_grpc.BlogServiceStub, blog_ids: List[str]):
    for id in blog_ids:
        blog_id = blog_pb2.BlogId(id=id)
        print("\nAsking for Blog with id %s" % (id))
        blog_response = stub.GetBlogInfo(blog_id)
        print("\nBlog with id %s, name %s and author %s, was obtained." %
              (blog_response.id, blog_response.name, blog_response.author))


def create_blog_posts(stub: blog_pb2_grpc.BlogServiceStub, blog_ids: List[str]) -> List[str]:
    posts_ids: List[str] = []
    for id in blog_ids:
        blog_post_creation = blog_pb2.BlogPostCreation(
            blog_id=id, post_title=gen.sentence(), post_content=gen.paragraph(min_sentences=3))

        blog_post_response = stub.CreateBlogPost(blog_post_creation)
        print("\nBlog Post with id %s, title %s and content %s, was created." % (
            blog_post_response.id, blog_post_response.title, blog_post_response.content))
        posts_ids.append(blog_post_response.id)
    return posts_ids


def get_blog_posts(stub: blog_pb2_grpc.BlogServiceStub, blog_ids: List[str]) -> List[Tuple[str, str]]:
    posts_id_content: List[Tuple[str, str]] = []
    for id in blog_ids:
        blog_id = blog_pb2.BlogId(id=id)
        print("\nAsking for Blog Posts of Blog with id %s" % (id))
        blog_post_responses = stub.GetBlogPosts(blog_id)
        for post in blog_post_responses:
            print("\nPost %s \n\nTitle: %s \n\nContent: %s" %
                  (post.id, post.title, post.content))
            posts_id_content.append((post.id, post.content))
    return posts_id_content


def record_quotes_from_posts(stub: blog_pb2_grpc.BlogServiceStub, posts_id_content: List[Tuple[str, str]]):
    quote_iterator = generate_quotes(posts_id_content)
    quote_summary = stub.RecordQuotes(quote_iterator)
    print("Quote Summary of quotes sent:\n")
    print(quote_summary)


def generate_quotes(posts_id_content: List[Tuple[str, str]]):
    for id, content in posts_id_content:
        words = list(map(str, content.split()))
        n_words_of_quote = random.randint(2, 5)
        start_of_quote = random.randint(0, len(words) - n_words_of_quote)
        quote_content = words[start_of_quote:start_of_quote + n_words_of_quote]
        quote = blog_pb2.Quote(content=' '.join(
            quote_content), blog_post_id=id)
        print("\nSending Quote %s of post %s" %
              (quote.content, quote.blog_post_id))
        yield quote
        time.sleep(random.randint(1, 3))


def send_blog_post_comments_to_chat(stub: blog_pb2_grpc.BlogServiceStub, post_id: str):
    responses = stub.BlogPostChat(generate_comments(post_id))
    for response in responses:
        print("\n\nReceived comment of blog %s:\nDate: %s\nAuthor: %s\nContent: %s" % (
            response.blog_post_id, datetime.datetime.fromtimestamp(response.created_at.seconds), response.author, response.content))


def generate_comments(post_id: str):
    author = gen.name()
    for _ in range(random.randint(2, 10)):
        yield blog_pb2.Comment(
            blog_post_id=post_id,
            author=author,
            content=gen.sentence(),
        )
        time.sleep(random.randint(1, 3))


def run():
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = blog_pb2_grpc.BlogServiceStub(channel)
        print("\n\n-------------- CreateBlog --------------")
        blog_ids = create_blogs(stub)
        print("\n\n-------------- GetBlogInfo --------------")
        get_blog_info(stub, blog_ids)
        print("\n\n-------------- CreateBlogPost --------------")
        posts_ids = create_blog_posts(stub, blog_ids)
        print("\n\n-------------- GetBlogPosts --------------")
        posts_id_content = get_blog_posts(stub, blog_ids)
        print("\n\n-------------- RecordQuotes --------------")
        record_quotes_from_posts(stub, posts_id_content)
        print("\n\n-------------- BlogPostChat --------------")
        send_blog_post_comments_to_chat(stub, random.choice(posts_ids))


if __name__ == '__main__':
    logging.basicConfig()
    run()

File: java-exercise/src/main/java/pt/up/blog/BlogClient.java
package pt.up.blog;

import io.grpc.Channel;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

public class BlogClient {

    /**
     * It is highly recommended to complete this exercise
     * having has support the following tutorial, please read it and have fun!
     * https://grpc.io/docs/languages/java/basics/
     */
    private final BlogServiceGrpc.BlogServiceBlockingStub blockingStub;
    // TODO
//    private final BlogServiceGrpc.BlogServiceStub asyncStub;
    private static final Logger logger = Logger.getLogger(BlogClient.class.getName());
    public BlogClient(Channel channel) {
        // create blocking stub
        blockingStub = BlogServiceGrpc.newBlockingStub(channel);
        // TODO create asynchronous stub
//        asyncStub =
    }

    /**
     * Creates 3 blogs with random information
     * @return an array with the ids of the generated blogs
     */
    private List<String> createBlogs() {
        List<String> blogIds = new ArrayList<>();

        for (int i = 0; i < 3; i++) {
            BlogCreation blogCreation = BlogCreation.newBuilder()
                    .setAuthor(generateRandomString(8))
                    .setName(generateRandomString(8))
                    .build();

            Blog blogResponse = blockingStub.createBlog(blogCreation);
            logger.info(String.format(
                    "Blog with id %s, name %s and author %s, was created.",
                    blogResponse.getId(),
                    blogResponse.getName(),
                    blogResponse.getAuthor()
            ));
            System.out.println();

            blogIds.add(blogResponse.getId());
        }

        return blogIds;
    }

    /**
     * Retrieve Blog information
     * @param blogIds list with ids of the desired blogs
     */
    private void getBlogInfo(List<String> blogIds) {
        for (String id: blogIds) {
            logger.info("Asking for Blog with id " + id);

            Blog blogResponse = blockingStub.getBlogInfo(
                    BlogId.newBuilder().setId(id).build()
            );

            logger.info(String.format(
                    "Blog with id %s, name %s and author %s, was obtained.",
                    blogResponse.getId(),
                    blogResponse.getName(),
                    blogResponse.getAuthor()
            ));
            System.out.println();
        }
    }

    /**
     * Creates posts on a specific blog
     * @param blogIds list with blog ids
     * @return list with the ids of posts created
     */
    private List<String> createBlogPosts(List<String> blogIds) {
        List<String> postsIds = new ArrayList<>();
        BlogPost blogPostResponse;
        BlogPostCreation blogPost;

        for (String id: blogIds) {

            /**
             * TODO 0
             * Create a Post instance that allows to call the server
             * and create a post on the given Blog
             *
             * It might help to check the .proto file for more information on which
             * fields belong to this service
             */
            blogPost = BlogPostCreation.getDefaultInstance(); // this code does not work

            blogPostResponse = blockingStub.createBlogPost(blogPost);
            logger.info(String.format(
                    "Blog Post with id %s, title %s and content %s, was created.",
                    blogPostResponse.getId(),
                    blogPostResponse.getTitle(),
                    blogPostResponse.getContent()
            ));
            System.out.println();

            postsIds.add(blogPostResponse.getId());
        }

        return postsIds;
    }

    /**
     * Retrieves a stream of blog posts
     * @param blogIds list with blog ids
     * @return list with post ids
     */
    private List<String> getBlogPosts(List<String> blogIds) {
        List<String> postsIds = new ArrayList<>();

        /**
         * TODO 1
         * This method allows to contact a service in which the server
         * sends a stream of messages.
         * In the next lines develop the necessary code to retrieve
         * the stream from the server.
         */
        // TODO 1 Here create a reference to Iterate through the received blog posts
        // Like the following one Iterator<String> aStringIterator
        // ? blogPostResponseIterator
        BlogPost blogPost;

        for (String id: blogIds) {
            // TODO 1 Add the appropriate information inside the next method call
//            blogPostResponseIterator = blockingStub.getBlogPosts();

            // TODO 2 When the tasks above are completed uncomment the next block of code
            /**
            while (blogPostResponseIterator.hasNext()) {
                blogPost = blogPostResponseIterator.next();

                logger.info(String.format(
                        "Post %s\nTitle: %s \nContent: %s",
                        blogPost.getId(),
                        blogPost.getTitle(),
                        blogPost.getContent()
                ));
                System.out.println();

                postsIds.add(blogPost.getId());
            }
             */
        }

        return postsIds;
    }

    /**
     * Sends a stream of Quotes
     * @param postIds lists of post ids
     * @throws InterruptedException
     */
    private void recordQuotesFromPosts(List<String> postIds) throws InterruptedException {
        final CountDownLatch finishLatch = new CountDownLatch(1);

        // This observer checks for whatever the server answers
        StreamObserver<QuoteSummary> responseObserver = new StreamObserver<QuoteSummary>() {
            @Override
            public void onNext(QuoteSummary summary) {
                /**
                 * As the server only sends a message (when all quotes were sent),
                 * then we can assume that this method will only be called one time with the
                 * quote summary
                 */
                logger.info(String.format(
                        "All quotes were sent. Yay!!"
                ));
                logger.info(String.format(
                        "In total were sent %d quotes relative to %d posts.",
                        summary.getQuoteCount(),
                        summary.getPostsCount()
                ));
            }

            @Override
            public void onError(Throwable t) {
                Status status = Status.fromThrowable(t);
                logger.warning(String.format("Record quotes Failed: %s", status.toString()));
                finishLatch.countDown();
            }

            @Override
            public void onCompleted() {
                logger.info("Finished Recording Quotes");
                finishLatch.countDown();
            }
        };

        /**
         * TODO 3 This is an asynchronous call so please go to the constructor
         * of this class and create an async stub
         */
//        StreamObserver<Quote> requestObserver = asyncStub.recordQuotes(responseObserver);

        Quote quote;
        Random rand = new Random();
        try {

            for(String id: postIds) {
                /**
                 * TODO 3 Build quotes and send them to the server. Just build, the send we already took care of below ;)
                 * The server will be waiting for quotes
                 * until the client calls the onCompleted() method
                 */
//                quote = Quote.newBuilder()

                // TODO 3 When the tasks above are completed uncomment the next block of code
                /**
                logger.info(String.format(
                        "Sending quote %s of post %s",
                        quote.getContent(),
                        quote.getBlogPostId()
                ));

                requestObserver.onNext(quote);
                 */

                // Sleep for a bit before sending the next one.
                Thread.sleep(rand.nextInt(1000) + 500);

                if (finishLatch.getCount() == 0) {
                    // RPC completed or errored before we finished sending.
                    // Sending further requests won't error, but they will just be thrown away.
                    return;
                }
            }
        } catch (RuntimeException e) {
            // TODO 3 also to uncomment
//            requestObserver.onError(e);
            throw e;
        }

        // TODO you already know what to do
//        requestObserver.onCompleted();

        finishLatch.await(1, TimeUnit.MINUTES);
    }

    /**
     * Simulates a chat where client and server send each other
     * messages continuously
     * @param postId a list of post ids
     * @throws InterruptedException
     */
    private void sendBlogPostCommentsToChat(String postId) throws InterruptedException {
        final CountDownLatch finishLatch = new CountDownLatch(1);


        /**
         * TODO 4 For the next line to work it needs a response observer that is triggered each time the
         * server sends back a comment
         *
         * Implement the appropriate observer and add him into the method call
         */
//        StreamObserver<Comment> requestObserver = asyncStub.blogPostChat(The observer is going to be here);

        Comment comment;
        Random rand = new Random();
        try {
            for (int i = 0; i < 2; i++) {
                comment = Comment.newBuilder()
                        .setBlogPostId(postId)
                        .setAuthor(generateRandomString(10))
                        .setContent(generateRandomString(20))
                        .build();

                logger.info(String.format(
                        "Sending comment with id %s and author %s",
                        comment.getBlogPostId(),
                        comment.getAuthor()
                ));

                // TODO 4 uncomment
//                requestObserver.onNext(comment);

                // Sleep for a bit before sending the next one.
                Thread.sleep(rand.nextInt(1000) + 500);
            }
        } catch (RuntimeException e) {
            // TODO 4 uncomment
//            requestObserver.onError(e);
            throw e;
        }

        // TODO 4 and uncomment one last time
//        requestObserver.onCompleted();

        finishLatch.await(1, TimeUnit.MINUTES);
    }

    // utils
    private String generateRandomString(int length) {
        int leftLimit = 97; // letter 'a'
        int rightLimit = 122; // letter 'z'
        Random random = new Random();

        return random.ints(leftLimit, rightLimit + 1)
          .limit(length)
          .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
          .toString();
    }

    public static void main(String[] args) throws Exception {
        String target = "localhost:50051";

        ManagedChannel channel = ManagedChannelBuilder.forTarget(target)
                // Channels are secure by default (via SSL/TLS). For the example we disable TLS to avoid
                // needing certificates.
                .usePlaintext()
                .build();
        try {
            BlogClient client = new BlogClient(channel);
            System.out.println("-------------- CreateBlog --------------");
            List<String> blogIds = client.createBlogs();

            System.out.println("-------------- GetBlogInfo --------------");
            client.getBlogInfo(blogIds);

            // TODO 1 uncomment
//            System.out.println("-------------- CreateBlogPost --------------");
//            List<String> postIds = client.createBlogPosts(blogIds);

            // TODO 2 uncomment
//            System.out.println("-------------- GetBlogPosts --------------");
//            client.getBlogPosts(blogIds);

            // TODO 3 uncomment
//            System.out.println("-------------- RecordQuotes --------------");
//            client.recordQuotesFromPosts(postIds);

            // TODO 4 uncomment
//            System.out.println("-------------- BlogPostChat --------------");
//            Random rand = new Random();
//            client.sendBlogPostCommentsToChat(postIds.get(rand.nextInt(postIds.size())));

            // TODO 4 uncomment the next line for bonus
//            System.out.println("\n-------------- Hope you enjoyed our little exercise. Until next time <3 --------------");
        } finally {
            // ManagedChannels use resources like threads and TCP connections. To prevent leaking these
            // resources the channel should be shut down when it will no longer be used. If it may be used
            // again leave it running.
            channel.shutdownNow().awaitTermination(5, TimeUnit.SECONDS);
        }
    }
}
